autoprecompiles/src/range_constraint_optimizer.rs

```diff
@@ -0,0 +1,146 @@
+use std::collections::BTreeMap;
+use std::fmt::Display;
+use std::hash::Hash;
+
+use num_traits::One;
+use powdr_constraint_solver::constraint_system::{
+    BusInteraction, BusInteractionHandler, ConstraintSystem,
+};
+use powdr_constraint_solver::grouped_expression::GroupedExpression;
+use powdr_constraint_solver::inliner::DegreeBound;
+use powdr_constraint_solver::range_constraint::RangeConstraint;
+use powdr_constraint_solver::solver::{new_solver, Solver};
+use powdr_number::FieldElement;
+
+pub type RangeConstraintMap<T, V> = BTreeMap<GroupedExpression<T, V>, RangeConstraint<T>>;
+
+pub trait RangeConstraintHandler<T: FieldElement> {
+    /// If the bus interaction *only* enforces range constraints, returns them
+    /// as a map of expressions to range constraints.
+    /// For example, if a bus interaction takes two arguments `a` and `b` and
+    /// enforces the range constraints `0 <= a < 2^b`, it is *not* a pure range
+    /// constraint if both values are unknown (because the allowed values of `a`
+    /// depend on `b`), but it is a pure range constraint if `b` is known.
+    /// Any stateful bus interaction is not a pure range constraint.
+    fn pure_range_constraints<V: Ord + Clone + Eq + Display + Hash>(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> Option<RangeConstraintMap<T, V>>;
+
+    /// Given a set of range constraints, returns a list of bus interactions
+    /// that implements them. The implementation is free to implement multiple
```

Comment:
What does it mean for a bus interaction to implement a range constraint? Is it that the set of bus interactions ensure those range constraints?